components/director/internal/domain/formationassignment/service.go

```diff
@@ -1081,21 +1047,108 @@ type AssignmentErrorWrapper struct {
 	Error AssignmentError `json:"error"`
 }
 
-func newNotificationStatusReportFromWebhookResponse(response *webhookdir.Response) *statusreport.NotificationStatusReport {
+func newNotificationStatusReportFromWebhookResponse(response *webhookdir.Response, operation model.FormationOperation, webhookMode graphql.WebhookMode) *statusreport.NotificationStatusReport {
 	var configuration json.RawMessage
 	if response.Config != nil {
 		configuration = []byte(*response.Config)
 	}
 
-	var state string
-	if response.State != nil {
-		state = *response.State
-	}
-
 	var errorFromResponse string
 	if response.Error != nil && *response.Error != "" {
 		errorFromResponse = *response.Error
 	}
 
-	return statusreport.NewNotificationStatusReport(configuration, state, errorFromResponse)
+	return statusreport.NewNotificationStatusReport(configuration, calculateState(response, operation, webhookMode), errorFromResponse)
+}
+
+func calculateState(response *webhookdir.Response, operation model.FormationOperation, webhookMode graphql.WebhookMode) string {
+	if response.State != nil && *response.State != "" {
+		return *response.State
+	}
+
+	if response.Error != nil && *response.Error != "" {
+		if operation == model.AssignFormation {
+			return string(model.CreateErrorAssignmentState)
+		}
+
+		return string(model.DeleteErrorAssignmentState)
+	}
+
+	if *response.ActualStatusCode == *response.SuccessStatusCode && webhookMode != graphql.WebhookModeAsyncCallback {
+		return string(model.ReadyAssignmentState)
+	}
+	if operation == model.AssignFormation && webhookMode == graphql.WebhookModeAsyncCallback {
+		return string(model.InitialAssignmentState)
+	}
+
+	if operation == model.AssignFormation && webhookMode == graphql.WebhookModeSync {
+		return string(model.ConfigPendingAssignmentState)
+	}
+	return string(model.DeletingAssignmentState)
```

Comment:
Maybe add a comment that this state shouldn't reach `Update` or `UpdateWithConstraints`? And that it only should happen when `operation == model.UnassignFormation && webhookMode == graphqlWebhookModeAsyncCallback` with a comment?

Reply:
The DELETING state actually reaches the UpdateWithConstraints in case of unassign and it should reach it so that we have recorded that the FA is being deleted